### PR TITLE
Bypass GetResponseXML to better match spec

### PR DESCRIPTION
**Pull request by liviuba**
Resolves https://github.com/servo/servo/issues/9553

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/9610)

### CODE REVIEW

**Comment by KiChjang** [comment]
Thanks for your pull request!

I believe @Ms2ger mentioned that he would like to move the caching in `GetResponseXML` into `document_response`, and looking at your PR, that change hasn't been made.

File: components/script/dom/xmlhttprequest.rs [hunk 1]
```diff
@@ -1068,6 +1067,7 @@ impl XMLHttpRequest {
             _ => { return None; }
         }
         temp_doc.set_encoding_name(DOMString::from(charset.name()));
+        self.response_xml.set(Some(temp_doc.r()));
```
Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: Unfortunately this still isn't the correct behavior yet; we need to check whether `self.response_xml` exists first before we commit to any sort of action - see steps [here](https://xhr.spec.whatwg.org/#document-response).

**Comment by Ms2ger** [comment]
Usually bors being happy would be good, but in this case it means I'll ask you to add some more tests, with `xhr.responseType` set to "document":

* getting `responseXML` twice in a row (this test should pass even with your changes)
* getting `response` twice in a row
* getting `response` followed by `responseXML`
* getting `responseXML` followed by `response`

In all cases, the returned documents should pass `assert_equals`. Please do ensure that the latter three tests fail with your changes.

See `tests/wpt/README.md` for more information about our testing infrastructure, and don't hesitate to come on IRC and ask questions if anything's unclear. (I'll be offline for the weekend myself, but other people should be able to help.)

See `tests/wpt/web-platform-tests/XMLHttpRequest/responsexml-basic.htm` for a

**Comment by liviuba** [comment]
Thanks for the feedback!

I'll write the tests/make sure my PR passes them soon

**Comment by Ms2ger** [comment]
I debugged this: the problem was that you were getting the responses before they were loaded, so you got two nulls in each case. This test catches the issue I found:
```js
      async_test(function() {
        var client = new XMLHttpRequest()
        client.open("GET", "resources/well-formed.xml")
        client.responseType = "document"
        assert_equals(client.responseType, "document")
        client.send()
        client.onload = this.step_func_done(function() {
          var first = client.response
          var second = client.response
          assert_not_equals(first, null)
          assert_not_equals(second, null)
          assert_equals(first, second)
        })
      }, "Getting response, then response")

      async_test(function() {
        var client = new XMLHttpRequest()
        client.open("GET", "resources/well-formed.xml")
        client.responseType = "document"
        assert_equals(client.responseType, "document")
        client.send()
        client.onload = this.step_func_done(function() {
          var first = client.responseXML
          var second = client.responseXML
          assert_not_equals(first, null)
          assert_not_equals(second, null)
          assert_equals(first, second)
        })
      }, "Getting responseXML, then responseXML")

      async_test(function() {
        var client = new XMLHttpRequest()
        client.open("GET", "resources/well-formed.xml")
        client.responseType = "document"
        assert_equals(client.responseType, "document")
        client.send()
        client.onload = this.step_func_done(function() {
          var first = client.responseXML
          var second = client.response
          assert_not_equals(first, null)
          assert_not_equals(second, null)
          assert_equals(first, second)
        })
      }, "Getting responseXML, then response")

      async_test(function() {
        var client = new XMLHttpRequest()
        client.open("GET", "resources/well-formed.xml")
        client.responseType = "document"
        assert_equals(client.responseType, "document")
        client.send()
        client.onload = this.step_func_done(function() {
          var first = client.response
          var second = client.responseXML
          assert_not_equals(first, null)
          assert_not_equals(second, null)
          assert_equals(first, second)
        })
      }, "Getting response, then responseXML")
```

Thanks again for working on this!

**Comment by KiChjang** [comment]
My comments still haven't been addressed.

**Comment by liviuba** [comment]
Yes, sorry, I'm on that now, first I wanted to fix a bad rebase.

File: components/script/dom/xmlhttprequest.rs (hunk 1 above)
Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: Btw, just implementing step 1 is good enough; I'll file an issue about implementing the rest of the steps correctly once this PR gets merged.

File: components/script/dom/xmlhttprequest.rs [hunk 2]
```diff
@@ -1126,36 +1126,53 @@ impl XMLHttpRequest {
...
-                if self.response_type.get() == XMLHttpRequestResponseType::_empty {
-                    return None;
-                }
-                else {
-                    temp_doc = self.document_text_html();
-                }
+        //step 1
+        match self.response_xml.get() {
+            Some(_) => {
+                return self.response_xml.get();
```
Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: `if self.response_xml.get().is_some() { return self.response_xml.get(); }`

File: components/script/dom/xmlhttprequest.rs [hunk 3]
```diff
@@ -1126,36 +1126,53 @@ impl XMLHttpRequest {
...
-        let temp_doc: Root<Document>;
-        match mime_type {
-            Some(Mime(mime::TopLevel::Text, mime::SubLevel::Html, _)) => {
-                if self.response_type.get() == XMLHttpRequestResponseType::_empty {
-                    return None;
-                }
-                else {
-                    temp_doc = self.document_text_html();
-                }
+        //step 1
```
Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: nit: Space after double slash and capitalize S.

File: components/script/dom/xmlhttprequest.rs [hunk 4]
```diff
@@ -1126,36 +1126,53 @@ impl XMLHttpRequest {
...
+                    },
+                    //step 4
+                    _ => { return None; }
                 }
-            },
-            _ => { return None; }
+                //step 9
+                temp_doc.set_encoding_name(DOMString::from(charset.name()));
+
+                //cache result
```
Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: This is step 13.

File: components/script/dom/xmlhttprequest.rs [hunk 5]
```diff
@@ -1126,36 +1126,53 @@ impl XMLHttpRequest {
...
-                }
-                else {
-                    return None;
+                let mime_type = self.final_mime_type();
+                //TODO: prescan the response to determine encoding if final charset is null
+                let charset = self.final_charset().unwrap_or(UTF_8);
+                let temp_doc: Root<Document>;
+                match mime_type {
+                    Some(Mime(mime::TopLevel::Text, mime::SubLevel::Html, _)) => {
+                        //step 5
```
Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: nit: Space after double slash and capitalize S.

File: components/script/dom/xmlhttprequest.rs [hunk 6]
```diff
@@ -1126,36 +1126,53 @@ impl XMLHttpRequest {
...
-                temp_doc = self.handle_xml();
-            },
-            Some(Mime(_, mime::SubLevel::Ext(sub), _)) => {
-                if sub.ends_with("+xml") {
-                    temp_doc = self.handle_xml();
-                }
-                else {
-                    return None;
+                let mime_type = self.final_mime_type();
+                //TODO: prescan the response to determine encoding if final charset is null
```
Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: nit: space after double slash.

File: components/script/dom/xmlhttprequest.rs [hunk 7]
```diff
@@ -1126,36 +1126,53 @@ impl XMLHttpRequest {
...
+                let charset = self.final_charset().unwrap_or(UTF_8);
+                let temp_doc: Root<Document>;
+                match mime_type {
+                    Some(Mime(mime::TopLevel::Text, mime::SubLevel::Html, _)) => {
+                        //step 5
+                        if self.response_type.get() != XMLHttpRequestResponseType::_empty {
+                            return None;
+                        }
+                        //step 6
```
Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: nit: Space after double slash and capitalize S.

File: components/script/dom/xmlhttprequest.rs [hunk 8]
```diff
@@ -1126,36 +1126,53 @@ impl XMLHttpRequest {
...
+                        else {
+                            temp_doc = self.document_text_html();
+                        }
+                    },
+                    //step 7
```
Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: nit: Space after double slash and capitalize S.

File: components/script/dom/xmlhttprequest.rs [hunk 9]
```diff
@@ -1126,36 +1126,53 @@ impl XMLHttpRequest {
...
+                    },
+                    Some(Mime(_, mime::SubLevel::Ext(sub), _)) => {
+                        if sub.ends_with("+xml") {
+                            temp_doc = self.handle_xml();
+                        }
+                        else {
+                            return None;
+                        }
+                    },
+                    //step 4
```
Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: nit: Space after double slash and capitalize S.

File: components/script/dom/xmlhttprequest.rs [hunk 10]
```diff
@@ -1126,36 +1126,53 @@ impl XMLHttpRequest {
...
+                    _ => { return None; }
                 }
-            },
-            _ => { return None; }
+                //step 9
```
Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: nit: Space after double slash and capitalize S.

File: components/script/dom/xmlhttprequest.rs [hunk 11]
```diff
@@ -1126,19 +1126,28 @@ impl XMLHttpRequest {
...
         let mime_type = self.final_mime_type();
-        //TODO: prescan the response to determine encoding if final charset is null
+        // TODO: prescan the response to determine encoding if final charset is null
         let charset = self.final_charset().unwrap_or(UTF_8);
         let temp_doc: Root<Document>;
         match mime_type {
             Some(Mime(mime::TopLevel::Text, mime::SubLevel::Html, _)) => {
-                if self.response_type.get() == XMLHttpRequestResponseType::_empty {
+                // Step 5
+                if self.response_type.get() != XMLHttpRequestResponseType::_empty {
```
Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: Why is this changed from `==` to `!=`?

File: components/script/dom/xmlhttprequest.rs [hunk 12]
```diff
@@ -1152,10 +1161,14 @@ impl XMLHttpRequest {
                     return None;
                 }
             },
+            // Step 4
             _ => { return None; }
         }
+        // Step 13
```
Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: This is step 9.

File: components/script/dom/xmlhttprequest.rs [hunk 13]
```diff
@@ -1152,10 +1161,14 @@ impl XMLHttpRequest {
...
         temp_doc.set_encoding_name(DOMString::from(charset.name()));
-        Some(temp_doc)
+        // Cache result
```
Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: This is step 13.

**Comment by KiChjang** [comment]
This is fine to merge after my nits and squashing.

**Comment by KiChjang** [comment]
Thanks for your contribution!

@bors-servo r+
